xbean-finder/src/main/java/org/apache/xbean/finder/archive/JarArchive.java

```diff
@@ -41,22 +44,63 @@ public class JarArchive implements Archive, AutoCloseable {
     private final JarFile jar;
     private final MJarSupport mjar = new MJarSupport();
 
-    public JarArchive(ClassLoader loader, URL url) {
+    /*
+     * Supports only 'file:/...' or 'jar:file:/...!/' URLs
+     */
+    public JarArchive(ClassLoader loader, URL url){
 //        if (!"jar".equals(url.getProtocol())) throw new IllegalArgumentException("not a jar url: " + url);
 
-        try {
-            this.loader = loader;
-            this.url = url;
-            URL u = url;
+        this.loader = loader;
+        this.url = url;
+        File jarFile = null;
+        String jarPath;
+        int idx;
 
-            String jarPath = url.getFile();
-            if (jarPath.contains("!")) {
-                jarPath = jarPath.substring(0, jarPath.indexOf("!"));
-                u = new URL(jarPath);
+        // Wipe out 'jar:' prefix AND '!/{...}' suffix(if any)
+        if("jar".equalsIgnoreCase(url.getProtocol())){
+
+            try{
+                jarPath = url.getPath();
+                url = new URL(jarPath.substring(0, jarPath.lastIndexOf("!/")));
+            }catch(MalformedURLException ex){
+                // Probably CPU overheat and/or DRAM undervoltage
```

Comment:
Is it really the cause or just the contract which enforces it so it is impossible to get?

Reply:
trying and trying - with no success... in THIS CASE i think just a contract

Reply:
Yes, it is really about the URL contract and potentially a wrong sub-protocol (not file nor a supported protocol)
Overall the point is that the comment is misleading so shouldn't be written this way and likely not rethrow it as a `UnsupportedOperationException` but more an `IllegalArgumentException` maybe, wdyt?

Reply:
i think since test for XBEAN-337 passes, it is time to close it as fixed. doing some cosmetic cleanup, if necessary

Reply:
yep but we don't want to get on main branch code with unexpected comments or exception so I think it is important to not only care of the passing case but more of error cases which are always the one costing the most ;)

Reply:
agree in general, and specifically what kind of RuntimeException it will i don't care